### ISSUE TITLE
Static linking is broken

### ISSUE THREAD

**Issue by rgrinberg** [janestreet/async_ssl#8]
fbc2610229ee1a203ec2d9302f3c188ee591ca0e breaks static linking openssl because these symbols are now being loaded dynamically.

Here's the kind of stacktrace you'd get on linux without openssl installed:

```
      ("Dl.DL_error(\"email_candidate_job.opt: undefined symbol: TLSv1_method\")"))
     (backtrace
      ("Raised at file \"src/memo.ml\", line 9, characters 22-23"
       "Called from file \"src/ssl.ml\", line 420, characters 2-14"
       "Called from file \"src/ssl.ml\", line 435, characters 4-40"
       "Called from file \"src/monitor.ml\", line 210, characters 25-32"
       "Called from file \"src/job_queue.ml\", line 164, characters 6-47" ""))
```

**Comment by diml** [comment]
What is the expected behavior in this case? Should it fail at compile time instead of run time?

**Comment by rgrinberg** [comment]
It shouldn't fail at all because the binary was compiled (and statically linked) on a linux box with libssl installed. Just to make it clear, here's the use case I have that's triggering this bug:

1) Compile binary on a linux box with open ssl
2) Make sure that c libs are statically linked (libssl in this case)
3) Deploy binary on another linux box
4) Current async_ssl gives me the error above. If I revert the commit mentioned above and repeat 1-3 then the binary works.

**Comment by trevorsummerssmith** [comment]
@diml I am curious if there's any movement on this, or there is anything we can do to help (submit a PR, etc). It would be really nice to have this functionality as it would allow us to use utop with async_ssl.

Thanks!

**Comment by diml** [comment]
Sorry for the delay, I think we looked into it but I can't remember what we concluded. A PR would be very welcome indeed.

I think we have to deal with this the hard way: at configure time, try to build C snippets to find out which methods are supported and generate a .h file containing:

```
#ifndef CONFIG_H
#define CONFIG_H
#define HAS_SSLv3_method
#undef HAS_TLSv1_method
...
#endif
```

then revert the commit pointed by @rgrinberg and in `ffi_bindings.ml`: 

```
#import "config.h"
...
#if defined(HAS_SSLv3_method)
let sslv3 = foreign "SSLv3_method" ssl_method_t
#else
let sslv3 () = failwith "not supported on this system"
#endif
```

You can look in core for examples of such scripts. Maybe ctypes could be used to detect what is supported, but I don't know enough about it to tell.

If you look into this, you should try to do the patch against the master branch. We'll then review it and I'll backport it to the 113.33 branch to do a minor release in opam.

**Comment by trevorsummerssmith** [comment]
Thanks for your response! This makes sense. I realized after posting this that I meant to post this on #10 not here. This is still very useful though, as the static linking is helpful for our deployment setup. I will look into getting a PR for this.

**Comment by diml** [comment]
I'm reviewing the latest changes before uploading them to github and I just saw that someone did this internally

**Comment by trevorsummerssmith** [comment]
Oh that's cool. So this is closed by the forthcoming release?

**Comment by diml** [comment]
Yeah, although it'll only be part of the development repository (janestreet/opam-repository) until the next stable release of Jane Street packages, unless we backport it to 113.33

**Comment by trevorsummerssmith** [comment]
Do you know when your next stable release will be? Thanks!

**Comment by diml** [comment]
Currently I'm working on getting jenga ready to build our public packages as maintaining separate build systems for 60+ packages is too much work. I think we'll do a stable release once this is done, I'd say in about 2 months

**Comment by diml** [comment]
In fact it was easy to backport the fix so I did it, I'll do a minor release tomorrow

**Comment by diml** [comment]
Submitted a minor release: ocaml/opam-repository#6720
